Sources/UrlFormEncoding/UrlFormDecoder.swift

```diff
@@ -0,0 +1,626 @@
+
+    public static let custom = ParsingStrategy.init
+  }
+}
+
+extension UrlFormDecoder.UnkeyedContainer.Key: CodingKey {
+  public var stringValue: String {
+    return String(self.index)
+  }
+
+  public init?(stringValue: String) {
+    guard let intValue = Int(stringValue) else { return nil }
+    self.init(intValue: intValue)
+  }
+
+  public var intValue: Int? {
+    return .some(self.index)
+  }
+
+  public init?(intValue: Int) {
+    self.init(index: intValue)
+  }
+}
+
+private let iso8601 = ((\DateFormatter.calendar) .~ Calendar(identifier: .iso8601))
+  >>> ((\DateFormatter.locale) .~ Locale(identifier: "en_US_POSIX"))
+  >>> ((\DateFormatter.timeZone) .~ TimeZone(abbreviation: "UTC"))
+
+private let iso8601DateFormatter = DateFormatter()
+  |> iso8601
+  |> \.dateFormat .~ "yyyy-MM-dd'T'HH:mm:ss.SSSXXXXX"
+
+private let iso8601DateFormatterWithoutMilliseconds = DateFormatter()
+  |> iso8601
+  |> \.dateFormat .~ "yyyy-MM-dd'T'HH:mm:ssXXXXX"
+
+private func parse(isArray: @escaping (String) -> Bool) -> (String) -> [String: Any] {
+  func parseHelp(_ params: inout [String: Any], _ path: [String], _ value: Any) {
+    let key = path[0]
+
+    if path.count == 1 {
+      params[key] = value
+    } else if path.count == 2 && isArray(path[1]) {
+      var values = params[key] as? [Any] ?? []
+      values.append(value)
+      params[key] = values
+    } else if isArray(path[1]) {
+      var (values, nested) = (params[key] as? [Any] ?? [], [:] as [String: Any])
+      parseHelp(&nested, Array(path[2...]), value)
+      values.append(nested)
+      params[key] = values
+    } else {
+      var values = params[key] as? [String: Any] ?? [:]
+      parseHelp(&values, Array(path[1...]), value)
+      params[key] = values
+    }
+  }
+
+  return { query in
+    var params: [String: Any] = [:]
+
+    for (name, value) in pairs(query) {
+      let result = name.reduce(into: (path: [] as [String], current: "")) { result, char in
+        switch char {
+        case "[":
+          if result.path.isEmpty {
+            result.path.append(result.current)
+            result.current.removeAll()
+          }
+        case "]":
+          result.path.append(result.current)
+          result.current.removeAll()
+        default:
+          result.current.append(char)
+        }
+      }
+      let path = result.current.isEmpty ? result.path : result.path + [result.current]
+      parseHelp(&params, path.isEmpty ? [""] : path, value)
+    }
+
+    return params
+  }
+}
+
+private func pairs(_ query: String) -> [(String, String)] {
+  return query
+    .split(separator: "&")
+    .map { (pairString: Substring) -> (name: String, value: String) in
+      let pairArray = pairString.split(separator: "=", maxSplits: 1, omittingEmptySubsequences: false)
+        .flatMap { String($0).removingPercentEncoding }
+      return (pairArray[0], pairArray.count == 2 ? pairArray[1] : "")
+    }
+    .sorted { $0.name < $1.name }
```

Comment:
if I remember correctly, `sorted` is not a stable sort, which means it's possible that the string `x[]=1&x[]=2` could get sorted into pairs `("x[]", "2"), ("x[]", "1")`. theoretically at least.

we can easily fix by doing `.enumerated().sorted().map(first)` or something. do you think it's worth it?

Reply:
Could we use our comparator for a stable sort? I'm surprised that sorting a list of people by their name may shuffle two "Margaret"s.

Reply:
> Could we use our comparator for a stable sort? 

Anything that uses `Array.sorted` under the hood (which our comparators do via helpers) is susceptible to unstable sorts. We can easily make a `stablySorted` that does the enumerated trick. And if we further have lenses on tuples, and comparators on lenses we could do something like (pseudo code):

```swift 
stablySorted = enumerated
  .sorted(second.comparator <> first.comparator)
  .map(second)
```